packages/nx/src/command-line/migrate.ts

```diff
@@ -12,92 +12,109 @@ import {
   writeJsonFile,
 } from '../utils/fileutils';
 import { logger } from '../utils/logger';
+import { NxMigrationsConfiguration, PackageJson } from '../utils/package-json';
 import {
-  checkForNPMRC,
-  detectPackageManager,
+  createTempNpmDirectory,
   getPackageManagerCommand,
   resolvePackageVersionUsingRegistry,
 } from '../utils/package-manager';
 import { handleErrors } from '../utils/params';
 
-type Dependencies = 'dependencies' | 'devDependencies';
+export type Dependencies = 'dependencies' | 'devDependencies';
 
-export type MigrationsJson = {
+export interface PackageJsonUpdateForPackage {
   version: string;
-  collection?: string;
-  generators?: {
-    [name: string]: { version: string; description?: string; cli?: string };
-  };
-  packageJsonUpdates?: {
-    [name: string]: {
-      version: string;
-      packages: {
-        [p: string]: {
-          version: string;
-          ifPackageInstalled?: string;
-          alwaysAddToPackageJson?: boolean;
-          addToPackageJson?: Dependencies;
-        };
-      };
+  ifPackageInstalled?: string;
+  alwaysAddToPackageJson?: boolean | Dependencies;
+  addToPackageJson?: boolean | Dependencies;
+}
+
+export type PackageJsonUpdates = {
+  [name: string]: {
+    version: string;
+    packages: {
+      [packageName: string]: PackageJsonUpdateForPackage;
     };
   };
 };
 
+export interface GeneratorMigration {
+  version: string;
+  description?: string;
+  cli?: string;
+  implementation?: string;
+  factory?: string;
+}
+
+export interface MigrationsJson {
+  version: string;
+  collection?: string;
+  generators?: { [name: string]: GeneratorMigration };
+  schematics?: { [name: string]: GeneratorMigration };
+  packageJsonUpdates?: PackageJsonUpdates;
+}
+
+export interface ResolvedMigrationConfiguration extends MigrationsJson {
+  packageGroup?: NxMigrationsConfiguration['packageGroup'];
+}
+
+const execAsync = promisify(exec);
+
 export function normalizeVersion(version: string) {
-  const [v, t] = version.split('-');
-  const [major, minor, patch] = v.split('.');
-  const newV = `${major || 0}.${minor || 0}.${patch || 0}`;
-  const newVersion = t ? `${newV}-${t}` : newV;
+  const [semver, prereleaseTag] = version.split('-');
+  const [major, minor, patch] = semver.split('.');
```

Comment:
The semver package could be used to parse these, rather than string manip. Would look something like this (but import at top of file instead of require and so on.

```suggestion
const {major, minor, patch, prerelease: [prereleaseTag, prereleaseNum]} = 
  require('semver')(version)
```